design/batch_retrieval_design.md

```diff
@@ -1,20 +1,215 @@
+
+| Status/Message   | Description                                                  |
+| ---------------- | ------------------------------------------------------------ |
+| 400 Bad Request  | Server cannot process request due to client error (invalid input parameters or logical validations). For example: adding `<>` in path to secret in Conjur/DAP |
+| 401 Unauthorized | The request lacks valid authentication credentials           |
+| 403 Forbidden    | The authenticated user lacks the necessary privilege         |
+| 404 Not Found    | Variable does not exist or variable does not have secret value |
+
+### Proposed API vs Current API
+
+|               | Current                                                      | Proposed                                                     |
+| ------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| URI           | `GET /secrets{?variable_ids}}`                               | `POST /secrets/batch/v2`                                     |
+| Request Body  | -                                                            | See **Request Body** section                                 |
+| Response Body | JSON of ***only*** successfully retrieved secrets and their values. Single failure, fails whole request and no secret is returned. | JSON of **all** secrets, their status code/messages, and values for successfully retrieved secrets |
+| Behavior      | Request fails when individual request fails                  | Request does not fail when sub/individual requests fail      |
+
+### Permissions
+
+Only an authorized Host with `read` and `execute` privileges on the Conjur/DAP secrets will be able to fetch the secret values. Otherwise, the proper 403 Forbidden error code will be returned for `status` and `value` of the response respectively. 
+
+During configuration of the Secrets Provider, the Host detailed for `authnLogin` is the Host that the Secrets Provider will use to authenticate to Conjur/DAP Follower and retrieve Conjur/DAP secrets. Therefore, this is the Host that will be used to make the request to this new Batch API endpoint and retrieve secrets from Conjur/DAP.
+
+### Mitigation
+
+Now that it is easy to brute-force multiple variables at a time in the body of the request, we should implement the necessary mitigations to close the attack surface.
+
+### Audit
+
+A new audit entry will made per secret fetched from Conjur/DAP as is done in the current Batch retrieval API endpoint.
+
+### Open questions
+
+1. Do we want this duplicated API documented? At this stage this API endpoint will be for internal use only.
+2. Do we ultimately want to move to a single Batch Request API and depricate the existing one?
+
+### Future enhancements
+
+1. Add secret versions as a filtering query parameter so we can get all secrets after a certain version
+2. Fetch all secrets that the user/host has permissions to
```

Comment:
This caught my eye...

As I understand it (correct me if I'm wrong), a given host in Conjur will always be used for a specific application (running on k8s).  If this is the case, and since we already have to configure Conjur (via policy) with the knowledge of what secrets a host can access, I wonder if the whole idea here might be redundant in a way....

That is, could we just have a `/host-secrets` endpoint that gets everything at once, since it already knows all the secrets the application will need?  Then the secrets-provider-for-k8s just sends its host id, gets back a structure with all the secrets, and loops through them, hydrating the k8s secrets with the mapped keys?

I guess what I'm asking is: Are we duplicating the knowledge of what secrets an application needs?  Putting that knowledge in both Conjur and in this batch request?

Reply:
Let me preface that this section is an accumulating of future intentions for this batch retrieval API endpoint that surpasses the work done in the secrets provider. We ultimately intend to make this endpoint forward facing and used across our teams. So for this point, I didn't think about this in the scope of the secrets provider.

> a given host in Conjur will always be used for a specific application (running on k8s)

A given host identity will be used by the secrets provider application container to fetch the secrets, not per customer application.

If we were to look at this idea in the scope of the secrets provider, it is not totally redundant because the customer may seek to get a secret from Conjur (by defining the path in its K8s Secret) that it actually doesn't have access to. Especially if they reuse a host for the secrets provider.

I am missing what the difference would be between the proposed batch extension idea (2️⃣ ) and `/host-secrets` here.

Reply:
Not totally sure I understood... Let me ask the question a different way:

Which piece of the system is responsible for holding the knowledge about which secrets an application has access to?  That is, I am a developer and write an application to be deployed on k8s needing secrets A, B, and C.  How does that information flow up the chain so that the k8s secrets provider:
1. Uses a host identity with access to those secrets.
2. Knows to request those secrets from Conjur and hydrate the corresponding k8 secrets values.

Reply:
> Which piece of the system is responsible for holding the knowledge about which secrets an application has access to?

Conjur. All defined in policy

1. The user defines the secrets they need A B C in the configuration of the secrets provider. For Helm it is done [here](https://github.com/cyberark/secrets-provider-for-k8s/blob/master/helm/secrets-provider/values.yaml#L26).
2. We ask the customer to add `conjur-map` key to the K8s Secrets A B C which holds the name of the individual secrets and their path in conjur. For example
```
apiVersion: v1
kind: Secret
metadata:
  name: test-k8s-secret
type: Opaque
stringData:
  conjur-map: |-
    secret: secrets/someSortOfSecret
    anotherSecret: secrets/anotherSomeSortOfSecret
```
3. As part of this configuration we also ask them to supply the `environment.conjur.authnLogin` [here](https://github.com/cyberark/secrets-provider-for-k8s/blob/master/helm/secrets-provider/values.yaml#L35) which Conjur will use to determine if host used has access to the secrets defined in the conjur-map path (from above example)
4. If so, we successfully update the K8s Secrets defined for us the k8sSecret parameter from step 1 with their values from Conjur.

Does that make it more sense?

Reply:
Starting to understand, but still a couple more questions.  Re: the above, and also your earlier comment:

> A given host identity will be used by the secrets provider application container to fetch the secrets, not per customer application.

it looks like we have:

1. A host, with access to a bunch of secrets.
2. Applications, each of which has access to some subset of those secrets.
3. A k8s secrets config (like the example in step 2. above) which defines the Conjur -> K8s name mappings for those secrets.  

Questions:

Re: 3, is there just a single file for the host?  Is there anything stopping one app from accessing k8s secrets that have been hydrated for another app?  If not, that seems like a security concern.  We want access restricted by application, but have to settle for the granularity of host level access, where a host has the authority to get the secrets for multiple applications?  Am I understanding that right?  Or is there a separate authnLogin for each application?

Reply:
>  is there just a single file for the host?

As in Conjur policy file? If so, yes. The user can define the level of granularity / resource restrictions in the scope of the namespace, service account, deployment (not for application containers b/c they run as a Job), etc. If I choose to restrict the secrets provider by service account, that means I am restricting all communication with Conjur at the level of the service account. In other words, only a Secrets Provider Pod that runs with that service account (and defined in policy) can authenticate/fetch secrets from Conjur.

>  Is there anything stopping one app from accessing k8s secrets that have been hydrated for another app?

In K8s, the K8s Secrets granularity is at the namespace level so all applications have access to all K8s secrets within the same namespace. So to answer your question, no. Our thinking was that the security measures taken within K8s environments is up to K8s to control. We restrict access and update K8s secrets with Conjur secrets that the Conjur host has access to. The rest is up for K8s to decide. 
The value in the Secrets Provider is its enhancement UX. The Secrets Provider allows customers to still use the K8s secrets struct but their secret's lifecycle can be managed in one place, Conjur.
Also, it is known that K8s secrets are not 💯 the best option for security. If the customer wants more security, usually they would look to summon etc.

> We want access restricted by application, but have to settle for the granularity of host level access, where a host has the authority to get the secrets for multiple applications?

Yes. I was to clarify, the host is only for the secrets provider, not for customer applications. If the customer wants to restrict access a bit more, they can deploy multiple secrets providers (application container solution - aka the work being done in this Milestone) or give the secrets provider that sits in the init containers different hosts for their secrets provider with different restrictions. But again, once it leaves our hands, it is up to K8s to enforce least privilege.

> Or is there a separate authnLogin for each application?

authnLogin is not for each customer application, rather for each secrets provider pod.

Reply:
Great explanation.  Thanks @sigalsax!

Final comments:

1. Returning to my original thought, now that I have a better understanding, it still seems to me like what we really want here is a single point, "get all my secrets as a batch," that an authenticated host can post to, especially since these hosts seem to exist for the sole purpose of identifying a k8s secrets provider.  It seems redundant that they have to specify the secrets they want in the request, when we can lookup and return all the secrets they own.

2. There may be some reason I'm not thinking of that prevents this suggestion, but.... it would be nice if even the mapping to k8s secrets was optional.  That is, by default, map to the same names that the secrets have in Conjur.  Now, I understand we may still want the list of secrets and their mapped values to exist within the k8s infrastructure in the file you pasted above:
```
apiVersion: v1
kind: Secret
metadata:
  name: test-k8s-secret
type: Opaque
stringData:
  conjur-map: |-
    secret: secrets/someSortOfSecret
    anotherSecret: secrets/anotherSomeSortOfSecret
```
since the admins managing k8s may be different from those managing Conjur.  However, we could still auto-generate the above file (using same name to same name mappings) from Conjur policy, in theory.  I think?  That way you'd have a single source of truth and cut down on the boilerplate files users have to write.

Reply:
My only concern with your point in 1 is we would be basically be dumping the Conjur DB. _Especially_ if the Conjur host is used for the secrets provider and reused for some other entity that require a different set of secrets. The consequence of that would be what I like to call "over sharing"; returning more secrets than asked for.

Regarding your second point, can you send me an example of the flow you are envisioning. I didn't quite understand
>  by default, map to the same names that the secrets have in Conjur.

@abrahamko FYI

Reply:
@jonahx i think that your suggestion for the host's secrets API is worth looking at but it can be added later. This design of @sigalsax and @abrahamko is for improving the current batch secrets retrieval (that will assist them in the secrets-provider). i think that both APIs can live together and have different value.

Reply:
@sigalsax @orenbm 

> My only concern with your point in 1 is we would be basically be dumping the Conjur DB. Especially if the Conjur host is used for the secrets provider and reused for some other entity that require a different set of secrets. 

I agree you don't want to overshare but shouldn't that problem be solved at the host level?  That is, to follow the principle of least privilege, you need to create hosts with properly scoped permissions.  If you create a profligate host and then reign it in with each request, you asking for trouble: every client/admin using the host has a new opportunity to fail and share something incorrectly.  Whereas if the host is created itself with just the permissions it needs, the task is done one time, by the Conjur admin, and then no one else can inadvertently mess it up.

> This design of @sigalsax and @abrahamko is for improving the current batch secrets retrieval (that will assist them in the secrets-provider)

Yeah I understand what I am saying is technically out of scope given the current proposal.  But I am suggesting that / asking if maybe the current proposal is not the ideal approach, both because of the repeated work and security concerns I mentioned.